mayastor/src/replicas/rebuild/rebuild_impl.rs

```diff
@@ -585,3 +462,175 @@ impl RebuildJob {
+    fn exec_internal_op(
+        &mut self,
+        op: RebuildOperation,
+    ) -> Result<(), RebuildError> {
+        self.exec_op(op, true)
+    }
+
+    /// Single state machine where all operations are handled
+    fn exec_op(
+        &mut self,
+        op: RebuildOperation,
+        override_pending: bool,
+    ) -> Result<(), RebuildError> {
+        type S = RebuildState;
+        let e = RebuildError::OpError {
+            operation: op.to_string(),
+            state: self.states.to_string(),
+        };
+
+        trace!(
+            "Executing operation {} with override {}",
+            op,
+            override_pending
+        );
+
+        match op {
+            RebuildOperation::Start => {
+                match self.state() {
+                    // start only allowed when... starting
+                    S::Stopped | S::Paused | S::Failed | S::Completed => Err(e),
+                    // for idempotence sake
+                    S::Running => Ok(()),
+                    S::Init => {
+                        self.states.set_pending(S::Running, false)?;
+                        self.schedule();
+                        Ok(())
+                    }
+                }
+            }
+            RebuildOperation::Stop => {
+                match self.state() {
+                    // We're already stopping anyway, so all is well
+                    S::Failed | S::Completed => Err(e),
+                    // for idempotence sake
+                    S::Stopped => Ok(()),
+                    S::Running => {
+                        self.states
+                            .set_pending(S::Stopped, override_pending)?;
+                        Ok(())
+                    }
+                    S::Init | S::Paused => {
+                        self.states
+                            .set_pending(S::Stopped, override_pending)?;
+
+                        // The rebuild is not running so we need to notify here
+                        self.notify();
+                        Ok(())
+                    }
+                }
+            }
+            RebuildOperation::Pause => match self.state() {
+                S::Stopped | S::Failed | S::Completed => Err(e),
+                S::Init | S::Running | S::Paused => {
+                    self.states.set_pending(S::Paused, false)?;
+                    Ok(())
+                }
+            },
+            RebuildOperation::Resume => match self.state() {
+                S::Init | S::Stopped | S::Failed | S::Completed => Err(e),
+                // for idempotence sake
+                S::Running => Ok(()),
+                S::Paused => {
+                    self.states.set_pending(S::Running, false)?;
+                    self.schedule();
+                    Ok(())
+                }
+            },
+            RebuildOperation::Fail => match self.state() {
+                S::Init | S::Stopped | S::Paused | S::Completed => Err(e),
+                S::Running | S::Failed => {
```

Comment:
Again, should the failed state be split out on its own?

Reply:
Done